ta/pkcs11/src/processing.c

```diff
@@ -1003,3 +1003,181 @@ enum pkcs11_rc entry_release_active_processing(struct pkcs11_client *client,
+
+	rc = serialargs_get_session_from_handle(&ctrlargs, client, &session);
+	if (rc)
+		return rc;
+
+	rc = serialargs_get(&ctrlargs, &wrapping_key_handle, sizeof(uint32_t));
+	if (rc)
+		return rc;
+
+	rc = serialargs_get(&ctrlargs, &key_handle, sizeof(uint32_t));
+	if (rc)
+		return rc;
+
+	rc = serialargs_alloc_get_one_attribute(&ctrlargs, &proc_params);
+	if (rc)
+		goto out;
+
+	if (serialargs_remaining_bytes(&ctrlargs)) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	rc = get_ready_session(session);
+	if (rc)
+		goto out;
+
+	wrapping_key = pkcs11_handle2object(wrapping_key_handle, session);
+	if (!wrapping_key) {
+		rc = PKCS11_CKR_WRAPPING_KEY_HANDLE_INVALID;
+		goto out;
+	}
+
+	key = pkcs11_handle2object(key_handle, session);
+	if (!key) {
+		rc = PKCS11_CKR_KEY_HANDLE_INVALID;
+		goto out;
+	}
+
+	/*
+	 * The wrapping key and key to be wrapped shouldn't be same.
+	 * PKCS#11 spec doesn't explicitly state that but logically this isn't
+	 * a use case and also acts as an attack vector, so explicitly
+	 * disallow this.
+	 */
+	if (key == wrapping_key) {
+		rc = PKCS11_CKR_WRAPPING_KEY_HANDLE_INVALID;
+		goto out;
+	}
+
+	rc = set_processing_state(session, function, wrapping_key, NULL);
+	if (rc)
+		goto out;
+
+	/* Check if mechanism can be used for wrapping function */
+	rc = check_mechanism_against_processing(session, proc_params->id,
+						function,
+						PKCS11_FUNC_STEP_INIT);
+	if (rc)
+		goto out;
+
+	/*
+	 * Check if wrapping key has CKA_WRAP set and its key type is
+	 * compatible with the mechanism passed
+	 */
+	rc = check_parent_attrs_against_processing(proc_params->id, function,
+						   wrapping_key->attributes);
+	if (rc) {
+		/*
+		 * CKR_KEY_FUNCTION_NOT_PERMITTED is not in the list of errors
+		 * specified with C_Wrap() in the specification. So
+		 * return the next most appropriate error.
+		 */
+		if (rc == PKCS11_CKR_KEY_FUNCTION_NOT_PERMITTED)
+			rc = PKCS11_CKR_WRAPPING_KEY_TYPE_INCONSISTENT;
+
+		goto out;
+	}
+
+	/* Check access of wrapping key */
+	rc = check_access_attrs_against_token(session,
+					      wrapping_key->attributes);
+	if (rc)
+		goto out;
+
+	switch (get_class(key->attributes)) {
+	case PKCS11_CKO_SECRET_KEY:
+		break;
+	/* Key type not supported as yet */
+	case PKCS11_CKO_PRIVATE_KEY:
+	default:
+		rc = PKCS11_CKR_KEY_NOT_WRAPPABLE;
+		goto out;
+	}
+
+	/* Check if key to be wrapped is extractable */
+	if (!get_bool(key->attributes, PKCS11_CKA_EXTRACTABLE)) {
+		DMSG("Extractable property is false");
+		rc = PKCS11_CKR_KEY_UNEXTRACTABLE;
+		goto out;
+	}
+
+	if (get_bool(key->attributes, PKCS11_CKA_WRAP_WITH_TRUSTED) &&
+	    !get_bool(wrapping_key->attributes, PKCS11_CKA_TRUSTED)) {
+		DMSG("Wrap with trusted not satisfied");
+		rc = PKCS11_CKR_KEY_NOT_WRAPPABLE;
+		goto out;
+	}
+
+	rc = check_access_attrs_against_token(session, key->attributes);
+	if (rc)
+		goto out;
+
+	rc = get_attribute_ptr(wrapping_key->attributes,
+			       PKCS11_CKA_WRAP_TEMPLATE, (void *)&req_attrs,
```

Comment:
is the cast needed?

Reply:
Compiler complains if cast is not put.

_ta/pkcs11/src/processing.c: In function ‘entry_wrap_key’:
ta/pkcs11/src/processing.c:1149:37: error: passing argument 3 of ‘get_attribute_ptr’ from incompatible pointer type [-Werror=incompatible-pointer-types]
 1149 |           PKCS11_CKA_WRAP_TEMPLATE, &req_attrs, &size);
      |                                     ^~~~~~~~~~
      |                                     |
      |                                     struct obj_attrs **
In file included from ta/pkcs11/src/processing.c:14:
ta/pkcs11/src/attributes.h:134:13: note: expected ‘void **’ but argument is of type ‘struct obj_attrs **’
  134 |      void **attr_ptr, uint32_t *attr_size);
      |      ~~~~~~~^~~~~~~~
cc1: all warnings being treated as errors_

Reply:
I suppose the right cast would then be `(void **)`

Reply:
Fixed.

Reply:
I was too quick to add the fix. This results in a compiler error - dereferencing type-punned pointer might break strict-aliasing rules [-Werror=strict-aliasing]

I think the best way to do this is to change the type of req_attrs to void * which will avoid all these casts. I will update the PR.